task/thermal/src/main.rs

```diff
@@ -12,111 +12,239 @@
+        Sensor {
+            device: Device::North(
+                Zone::Central,
+                Tmp116::new(&devices::tmp117_north(task)),
+            ),
+            id: sensors::TMP117_NORTH_TEMPERATURE_SENSOR,
+        },
+        Sensor {
+            device: Device::North(
+                Zone::West,
+                Tmp116::new(&devices::tmp117_northwest(task)),
+            ),
+            id: sensors::TMP117_NORTHWEST_TEMPERATURE_SENSOR,
+        },
+        Sensor {
+            device: Device::South(
+                Zone::East,
+                Tmp116::new(&devices::tmp117_southeast(task)),
+            ),
+            id: sensors::TMP117_SOUTHEAST_TEMPERATURE_SENSOR,
+        },
+        Sensor {
+            device: Device::South(
+                Zone::Central,
+                Tmp116::new(&devices::tmp117_south(task)),
+            ),
+            id: sensors::TMP117_SOUTH_TEMPERATURE_SENSOR,
+        },
+        Sensor {
+            device: Device::South(
+                Zone::West,
+                Tmp116::new(&devices::tmp117_southwest(task)),
+            ),
+            id: sensors::TMP117_SOUTHWEST_TEMPERATURE_SENSOR,
+        },
+        Sensor {
+            device: Device::CPU(SbTsi::new(&devices::sbtsi(task)[0])),
+            id: sensors::SBTSI_TEMPERATURE_SENSOR,
+        },
+    ]
+}
+
+struct ServerImpl {
+    sensor: sensor_api::Sensor,
+    sensors: [Sensor; NUM_TEMPERATURE_SENSORS],
+    fctrl: Max31790,
+    deadline: u64,
+}
+
+const TIMER_MASK: u32 = 1 << 0;
+const TIMER_INTERVAL: u64 = 1000;
+
+impl ServerImpl {
+    fn read_fans(&self) {
+        let ids = &sensors::MAX31790_SPEED_SENSORS;
+
+        for ndx in 0..MAX_FANS {
+            let fan = Fan::from(ndx);
+
+            match self.fctrl.fan_rpm(fan) {
+                Ok(reading) => {
+                    self.sensor
+                        .post(ids[ndx as usize], reading.0.into())
+                        .unwrap();
+                }
+                Err(e) => {
+                    self.sensor.nodata(ids[ndx as usize], e.into()).unwrap()
+                }
+            }
+        }
     }
 }
 
-fn temp_read<E: core::fmt::Debug, T: TempSensor<E> + core::fmt::Display>(
-    device: &T,
-) {
-    match device.read_temperature() {
-        Ok(temp) => {
-            print_temp(temp, device);
+impl idl::InOrderThermalImpl for ServerImpl {
+    fn set_fan_pwm(
+        &mut self,
+        _: &RecvMessage,
+        index: u8,
+        pwm: u8,
+    ) -> Result<(), RequestError<ThermalError>> {
+        if index < MAX_FANS {
+            let fan = Fan::from(index);
+
+            if pwm <= 100 {
+                match self.fctrl.set_pwm(fan, PWMDuty(pwm)) {
+                    Ok(_) => Ok(()),
+                    Err(_) => Err(ThermalError::DeviceError.into()),
+                }
+            } else {
+                Err(ThermalError::InvalidPWM.into())
+            }
+        } else {
+            Err(ThermalError::InvalidFan.into())
```

Comment:
FWIW, this and the error on line 181 look like client programming errors, which is a great use for `RequestError::BadMessage` (which will panic the client).

Reply:
In this case, the client is often `hiffy`, which I would rather not detonate. ;)